Megnha (#4)
bitprj/BitHeroesS20: Meghna/README.md

# Scheduling Bot

### Meghna Pillai

> Sometimes (well most of the time), humans tend to forget things. And while there may be several methods of scheduling reminders, having a chatbot is one of the most simple applications you can have to help with this. A scheduling chatbot can ask you several questions regarding the details of your event and puts it all together to send you a reminder one hour prior. In this tutorial, you will learn how to create a chatbot using Azure services that will make your life easier and up-to-date. There are three parts to this: developing the actual Bot, creating a web app, and creating an Azure function. Let's get started! 

Reviewer: ```suggestion
Sometimes (well, most of the time), humans tend to forget things. And while there may be several methods of scheduling reminders, having a chatbot is one of the simplest applications you can have to help out. A scheduling chatbot can ask you several questions regarding the details of your event and put it all together to send you a reminder one hour beforehand. In this tutorial, you will learn how to create a chatbot using Azure services that will make your life easier and up-to-date. There are three parts to this: developing the Bot's code, creating a web app, and creating an Azure function. Let's get started! 
```

5. On the `Identity verification by phone` page, choose your country code, and type in your phone number. Enter the verification code sent to you and select `Verify code`.
6. If the verification code is correct, it will prompt you to enter your credit card details. When done, select `Next`.
7. Review the agreement and privacy statement and select `Sign up`.

**Creating New Folders To Store Files** 

When you are all signed up for Azure, the next step is to create a few folders on your computer. 

1. Go to the local disk on your computer.

2. Create a folder. For our purposes, we will name the folder `Scheduling Bot`, but feel free to enter any name you prefer.

3. Once you create a folder, create three other folders within the first one. Feel free to name it anything you like, but for our purposes, we will call it the following: 

   `SchedulerBot (1)`

   `SchedulerBotWebApp (2)`

   `Scheduler Function (3)`

   ![](Screenshots/Screenshot%20(6).png)

**Composing The Bot (Stage 1)**

We must now start composing the scheduler bot using the Bot Framework Composer. 

> *Side note: The Bot that we are creating will not have any advanced features, such as intents or Triggers, as it will only extend the `WelcomeTheUser` activity. Regardless, it gets the main job done, and is a cool program to create! 

Reviewer: ```suggestion
*Side note: The Bot that we are creating will not have any advanced features, such as intents or Triggers, as it will only extend the `WelcomeTheUser` activity. Regardless, it gets the job done, is easy to extend, and is a cool program to create! 
```

Let's get started.

**Phase 1 - Set Up**

1. Go to Bot Framework Composer (one of the five programs that you downloaded on your computer). 

2. Select `New` and then `Create from scratch` when you get to the home page.

3. Under `Name`, give your Bot a name; it can be anything you would like. In this demo, we will call it `SchedulerBot`. And under `Location`, save your Bot to the first Bot folder that we made, which is inside the master folder that we also created. For instance, in this demo, the master folder is called `Scheduling Bot`, and we will be choosing our Location to be `SchedulerBot (1)`, the first of the three main folders inside the master. The other folders in the master are for the Web App and Function that we will be creating later on in this tutorial. 

   ![](Screenshots/Screenshot%20(5).png)

> 4. When redirected, if you look to the left, under `Filter Dialog`, there should be a section that says `Greeting`. Select that.

Reviewer: ```suggestion
4. When redirected, if you look to the left, under `Filter Dialog`, select the section that says `Greeting`. 
```

3. Next to `User Input`, there should be an option called `Other`. In the text box underneath `Unrecognized prompt`, we must type a message that the user will receive if the max turn count (if mentioned) has exceeded and the default value is selected as the value. For this demo, please enter the following:

   ```
   - Sorry I don't recognize ${this.value}. Please enter your name as text greater than 3 and less than 25 characters.
   ```

   The `${this.value}` will allow the Bot to respond to the user using the response they entered in the first place.

4. Moving along, under `Validation Rules`, we must create an expression that restricts the name that the user enters to be more than three characters and less than twenty-five and validates user input. Enter in the following to the right of `fx`:

   ```
   =length(this.value) > 3 && length(this.value) < 25
   ```

   Now, this will allow the Bot to only accept names that are greater than three and less than twenty-five characters.

   *Important note: Whenever you enter a validation rule, make sure to hit enter at the end so that it registers. Otherwise, it will not save.

> 5. When you scroll down the side bar on the right, there will be a title called `Invalid prompt`. In this text box, we must type a message for the Bot to send when the user input does not meet any validation expression. Enter the following:

Reviewer: ```suggestion
5. When you scroll down the sidebar on the right, there will be a title called `Invalid prompt`. In this text box, we must type a message for the Bot to send when the user input does not meet any validation expression. Enter the following:
```

3. Next to `User Input`, there should be an option called `Other`. In the text box underneath `Unrecognized prompt`, we must type a message that the user will receive if the max turn count (if mentioned) has exceeded and the default value is selected as the value. For this demo, please enter the following:

   ```
   - Please enter a valid ten-digit number.
   ```

4. When looking at `Validation Rules`, we must create an expression that restricts the phone number that the user enters to be ten characters and validates user input. Enter in the following to the right of `fx`:

   `=length(this.value)==10`

   Now, this will allow the Bot to accept characters that are only equal to ten characters. For this demo, we are only validating the length of the input, and not the content. 

> 5. When you scroll down the side bar on the right, there will be a title called `Invalid prompt`. In this text box, we must type a message for the Bot to send when the user input does not meet any validation expression. Enter the following:

Reviewer: ```suggestion
5. When you scroll down the sidebar on the right, there will be a title called `Invalid prompt`. In this text box, we must type a message for the Bot to send when the user input does not meet any validation expression. Enter the following:
```

   ```

>    - I don't recognized ${this.value} as a valid phone number.

Reviewer: ```suggestion
   - I don't recognize ${this.value} as a valid phone number.
```

3. Once you are logged in to your database, press `Ctrl + N` to open a new query. Enter the following code in your new query:

   ```
   if(OBJECT_ID('Schedule', 'U') is not null)
   drop table dbo.Schedule;
   
   create table dbo.Schedule
   (
   	ID int identity,
   	UserId nvarchar(256),
   	EventDate DateTime,
   	EventName varchar(256),
   	MinutesToAlertBefore int,
   	Completed bit,
   	PhoneNumber varchar(10),
   	CONSTRAINT [PK_Schedule] PRIMARY KEY CLUSTERED ([ID] ASC)
   )
   ```

   When describing the code, the first two lines delete the table if it already exists, and the rest of the lines create the table that we need to store user input.

4. Change the database at the top of the screen from `Master` to the one you created. Otherwise, the table will end up in the Master database instead of the one created for this Bot. In this demo, the database we created is labeled as `SchedulingBotDB`. 

5. Select all of the code and press `Execute`. What this code does is create a new table for you to store all the data that comes from the Bot when a user enters in the details regarding a reminder or event.

> 6. Press the blue refresh button towards the upper left of your screen. Once you refresh, you should be able to see that you created a table with several columns to store your all the relevant data from your Bot. 

Reviewer: ```suggestion
6. Press the blue refresh button towards the upper left of your screen. Once you refresh, you should be able to see that you created a table with several columns to store all the relevant data from your Bot. 
```

      a. Select your language as `C#`.
      b. Change your project type to `web`.
      c. Choose `ASP.NET Web Application (.NET Framework)` 
   	d. Hit  `Next`.

   ![](Screenshots/Screenshot%20(12).png)

2. Now, you must name the project. In this demo, we're going to name it `SchedulingBotWebApp`. Once named, make sure to save it to the Web App folder in your master folder (The names of the folders may be different depending on what you inputted originally). If you named the folders according to what the demo described, then you should save it to the `SchedulerBotWebApp (2)` folder in `Scheduling Bot`. 

*Installing Dapper*

Before we get started, we need to download a NuGet package necessary for this particular project. We will use `Dapper`, the NuGet package we that we are installing, to get the API to talk to the database, and also to convert the data retrieved from the database into a corresponding data model defined in the API.

> 1. To do this, right click on `SchedulingBotWebApp` and click on `Manage NuGet Packages...`. Here's a visual to help: 

Reviewer: ```suggestion
1. To do this, right-click on `SchedulingBotWebApp` and click on `Manage NuGet Packages...`. Here's a visual to help: 
```

   ![](Screenshots/Screenshot%20(23).png)

2. When redirected, select `Browse` and look up `Dapper`. Then click on it and follow the procedures for installation. 

*Creating a New Class in Models*

> 1. If you look to the right, you should see a `Solution Explorer`. Scroll down to find the `Models` folder, and right click on it. Then, hover over `Add` and select `Class...`. 

Reviewer: ```suggestion
1. If you look to the right, you should see a `Solution Explorer`. Scroll down to find the `Models` folder, and right-click on it. Then, hover over `Add` and select `Class...`. 
```

   ![](Screenshots/Screenshot%20(17).png)

   Rename the class as `Schedule` for now. 

2. When redirected to the new class you added, insert the following code into `public class Schedule {}`:

   ```
           public int ID { get; set; }
           public string UserId { get; set; }
           public DateTime EventDate { get; set; }
           public string EventName { get; set; }
           public int MinutesToAlertBefore { get; set; }
           public Boolean Completed { get; set; }
           public string PhoneNumber { get; set; }
   ```

   It should look like this: 

   ![](Screenshots/Screenshot%20(19).png)

   The code you just entered is the data structure of the table. It matches with what's in the database.  

*Creating a New Class in Database*

> 1. Let's create a new folder. In your `Solution Explorer`, right click on `SchedulingBotWebApp`. Then, hover over `Add` and select `New Folder`.  Name the folder `Database`. Here's a visual to help you do this: 

Reviewer: ```suggestion
1. Let's create a new folder. In your `Solution Explorer`, right-click on `SchedulingBotWebApp`. Then, hover over `Add` and select `New Folder`.  Name the folder `Database`. Here's a visual to help you do this: 
```

   ![](Screenshots/Screenshot%20(21).png)

> 2. Now, as we did before, we need to create a new class in this folder.  To do so, scroll down to find the `Database` folder, and right click on it. Then, hover over `Add` and select `Class...`. Rename the class as `Schedule`.  

Reviewer: ```suggestion
2. Now, as we did before, we need to create a new class in this folder.  To do so, scroll down to find the `Database` folder, and right-click on it. Then, hover over `Add` and select `Class...`. Rename the class as `Schedule`.  
```

               using (IDbConnection db = new SqlConnection(ConnectionString))
               {
                   db.Execute(ScheduleComplete, new { Id = schedule.ID }, commandType: CommandType.Text);
               }
           }
       }
   ```

4. Once you deploy the code, you may realize that a few lines have been underlined in red, indicating some sort of issue. Hover over each line with red underlined, press the light bulb symbol, and select the first suggestion that pops up to resolve the issues. Here's a visual to guide you: 

   ![](Screenshots/Screenshot%20(25).png)

5. Once you are done, there should no longer be any visible errors on your page. 

   What you just created is your data layer. It allows you to send data (by inserting a row into the table), read data, and update data in the database. We have created three methods to match those actions accordingly in the code you inserted. Your controller will end up calling this class that you created. To describe what a controller is, it is the front-end. Your Bot and your Azure Function, which we have not created yet, is what we will end up exposing your controller to.

*Creating a New Class in Controllers*

> 1. Scroll to find the `Controllers` folder, and right click on it. Then, hover over `Add` and select `Class...`. Name the class `SchedulingAPIController`. 

Reviewer: ```suggestion
1. Scroll to find the `Controllers` folder, and right-click on it. Then, hover over `Add` and select `Class...`. Name the class `SchedulingAPIController`. 
```

   The class we just created has three main functionalities as well: it saves the schedule, it gets schedules, and it also marks them as complete. Here you are calling the `Schedule.cs` class that you created within the database folder, which in turn calls the database using a Connection String.

*Editing `Web.config`*

1. Go back to your `Solution Explorer`. Scroll down to find `Web.config` and click on it. 

2. Immediately beneath ` </appSettings>`, enter the following code. Remember to replace `{your_connection_string}` with your own connection string that you took note of towards the beginning of this tutorial. 

   ```
   <connectionStrings>
       <add name="DefaultConnection" connectionString="{your_connection_string}" providerName="System.Data.SqlClient" />
     </connectionStrings>
   ```

   Finally, remember to save all of your work by clicking `File` and then `Save all`.

*Publishing Web App To Azure*

> 1. Now it's time for us to publish everything to Azure. To do this, right click on `SchedulingBotWebApp`, and select `Publish... ` . When it asks, `Where are you publishing today?`, choose `Azure`. And then, when it asks, `Which Azure service would you like to use to host your application?`, select `Azure App Service (Windows)`. 

Reviewer: ```suggestion
1. Now it's time for us to publish everything to Azure. To do this, right-click on `SchedulingBotWebApp`, and select `Publish... `. When it asks, `Where are you publishing today?`, choose `Azure`. And then, when it asks, `Which Azure service would you like to use to host your application?`, select `Azure App Service (Windows)`. 
```

      a. Select your language as `C#`.
      b. Change your platform to `Azure`. Change your project type to `web`.
   ​	c. Choose `Azure Functions`. 
   ​	d. Hit  `Next`.

   ![](Screenshots/Screenshot%20(30).png)

2. Now, you must name the project. In this demo, we're going to name it `SchedulingBotFunction`. Once named, make sure to save it to the Function in your master folder (The names of the folders may be different depending on what you inputted originally). If you named the folders according to what the demo described, then you should save it to the `SchedulerFunction (3)` folder in `Scheduling Bot`. 

   ![](Screenshots/Screenshot%20(31).png)

3. Select `Timer Trigger`. Change the 5 (runs every 5 minutes) to a 1 (runs every one minute), and press `Create`. 

*Installing Twilio*

1. Before we get started, we need to download a NuGet package necessary for this particular project. To do this, right click on `SchedulingBotWebApp` and click on `Manage NuGet Packages...`. Here's a visual to help: 

   ![](Screenshots/Screenshot%20(23).png)

2. When redirected, select `Browse` and look up `Twilio`. Then click on it and follow the procedures for installation. 

> *Set Up*

Reviewer: ```suggestion
*Set-Up*
```

   
                       TwilioClient.Init(accountSid, authToken);
   
                       var message = MessageResource.Create(
                           to: new PhoneNumber("schedule.PhoneNumber"),
                           from: new PhoneNumber("{your_Twilio_phone_number}"),
                           body: $"Reminder from SchedulerBot. {schedule.EventName} at {schedule.EventDate}");
   
                       string updateurl = $"{your_update_URL}";
                       var content = new StringContent(JsonConvert.SerializeObject(schedule), Encoding.UTF8, "application/json");
   
                       using (var _httpClient = new HttpClient())
                       {
                           var r = await _httpClient.PostAsync(updateurl, content);
                       }
   
                   }
               }
   ```

2. Once you deploy the code, you may realize that a few lines have been underlined in red, indicating some sort of issue. Hover over each line with red underlined, press the light bulb symbol, and select the first suggestion that pops up to resolve the issues. 

*Publishing Function To Azure*

> 1. Now it's time for us to publish everything to Azure. To do this, right click on `SchedulingBotFunction`, and select `Publish... ` . When it asks, `Where are you publishing today?`, choose `Azure`. And then, when it asks, `Which Azure service would you like to use to host your application?`, select `Azure Function App (Windows)`. 

Reviewer: ```suggestion
1. Now it's time for us to publish everything to Azure. To do this, right-click on `SchedulingBotFunction`, and select `Publish... `. When it asks, `Where are you publishing today?`, choose `Azure`. And then, when it asks, `Which Azure service would you like to use to host your application?`, select `Azure Function App (Windows)`. 
```
